src/Umbraco.Web/Unversion/UnversionService.cs

```diff
@@ -0,0 +1,169 @@
+            if (_config.ConfigEntries.ContainsKey(content.ContentType.Alias))
+                configEntries.AddRange(_config.ConfigEntries[content.ContentType.Alias]);
+
+            // Add all config entries (would it be more than one though?)
+            if (_config.ConfigEntries.ContainsKey(UnversionConfig.AllDocumentTypesKey))
+                configEntries.AddRange(_config.ConfigEntries[UnversionConfig.AllDocumentTypesKey]);
+
+            if (configEntries.Count <= 0)
+            {
+                // Would this ever get hit ?
+                _logger.Debug<UnversionService>("No unversion configuration found for type {ContentTypeAlias}", content.ContentType.Alias);
+                return;
+            }
+
+            foreach (var configEntry in configEntries)
+            {
+                var isValid = true;
+
+                // Check if RootXPath is configured
+                if (!string.IsNullOrEmpty(configEntry.RootXPath))
+                    if (content.Level > 1 && content.ParentId > 0)
+                    {
+                        var ids = GetNodeIdsFromXpath(configEntry.RootXPath);
+                        isValid = ids.Contains(content.ParentId);
+                    }
+
+                if (!isValid)
+                {
+                    _logger.Debug<UnversionService>("Invalid RootXPath configuration for '{ContentTypeAlias}' the XPath selector does not contain the parent id '{ContentParentId}' for the current content item '{ContentUdi}'.",
+                        configEntry.DocTypeAlias, content.ParentId, content.GetUdi());
+
+                    continue;
+                }
+
+                // Get every version of content node
+                var allVersions = _contentService.GetVersionsSlim(content.Id, 0, int.MaxValue).ToList();
+                if (!allVersions.Any())
+                {
+                    _logger.Debug<UnversionService>("No versions of content '{ContentUdi}' found", content.GetUdi());
+                    continue;
+                }
+
+                // Determine versions to delete
+                var versionIdsToDelete = GetVersionsToDelete(allVersions, configEntry, DateTime.Now);
+
+                // Delete versions
+                foreach (var vid in versionIdsToDelete)
+                {
+                    _logger.Info<UnversionService>("Deleting version {VersionId} of content '{ContentName}' {ContentUdi}", vid, content.Name, content.GetUdi());
+
+                    // Ensure we only delete the specified version and do NOT delete PRIOR versions
+                    _contentService.DeleteVersion(content.Id, vid, false);
+                }
+            }
+
+        }
+
+        /// <summary>
+        /// Iterates a list of IContent versions and returns items to be removed based on a configEntry.
+        /// </summary>
+        public List<int> GetVersionsToDelete(List<IContent> versions, UnversionConfigEntry configEntry, DateTime currentDateTime)
```

Comment:
List is mutable, only return mutable objects if you plan on allowing them to be mutable which is almost no cases for returning collections from methods. IEnumerable is not mutable.